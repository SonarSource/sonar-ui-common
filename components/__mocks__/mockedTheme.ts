/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

// IMPORTANT: any change in this file requires restart of the dev server
const grid = 8;

export const mockedTheme = {
  colors: {
    blue: '#4b9fd5',
    veryLightBlue: '#f2faff',
    lightBlue: '#cae3f2',
    darkBlue: '#236a97',
    veryDarkBlue: '#0E516F',
    green: '#00aa00',
    lightGreen: '#b0d513',
    veryLightGreen: '#f5f9fc',
    yellow: '#eabe06',
    orange: '#ed7d20',
    red: '#d4333f',
    purple: '#9139d4',

    gray94: '#efefef',
    gray80: '#cdcdcd',
    gray71: '#b4b4b4',
    gray67: '#aaa',
    gray60: '#999',
    gray40: '#404040',

    transparentWhite: 'rgba(255,255,255,0.62)',
    transparentGray: 'rgba(200, 200, 200, 0.5)',

    disableGrayText: '#bbb',
    disableGrayBorder: '#ddd',
    disableGrayBg: '#ebebeb',

    barBackgroundColor: '#f3f3f3',
    barBorderColor: '#e6e6e6',

    globalNavBarBg: '#262626',

    // fonts
    baseFontColor: '#444',
    secondFontColor: '#777',

    // forms
    mandatoryFieldColor: '#a4030f',

    // leak
    leakColor: '#fbf3d5',
    leakBorderColor: '#eae3c7',

    // issues
    issueBgColor: '#f2dede',
    hotspotBgColor: '#eeeff4',
    issueLocationSelected: '#f4b1b0',
    issueLocationHighlighted: '#e1e1f2',
    conciseIssueRed: '#d18582',
    conciseIssueRedSelected: '#a4030f',

    // codeviewer
    lineCoverageRed: '#a4030f',
    lineCoverageGreen: '#b4dd78',

    // alerts
    warningIconColor: '#e2bf41',

    alertBorderError: '#f4b1b0',
    alertBackgroundError: '#f2dede',
    alertTextError: '#862422',
    alertIconError: '#a4030f',

    alertBorderWarning: '#faebcc',
    alertBackgroundWarning: '#fcf8e3',
    alertTextWarning: '#6f4f17',
    alertIconWarning: '#db781a',

    alertBorderSuccess: '#d6e9c6',
    alertBackgroundSuccess: '#dff0d8',
    alertTextSuccess: '#215821',
    alertIconSuccess: '#6d9867',

    alertBorderInfo: '#b1dff3',
    alertBackgroundInfo: '#d9edf7',
    alertTextInfo: '#0e516f',
    alertIconInfo: '#0271b9'
  },

  sizes: {
    gridSize: `${grid}px`,

    baseFontSize: '13px',
    verySmallFontSize: '10px',
    smallFontSize: '12px',
    mediumFontSize: '14px',
    bigFontSize: '16px',
    hugeFontSize: '24px',

    hugeControlHeight: `${5 * grid}px`,
    largeControlHeight: `${4 * grid}px`,
    controlHeight: `${3 * grid}px`,
    smallControlHeight: `${2.5 * grid}px`,
    tinyControlHeight: `${2 * grid}px`,

    globalNavHeight: `${6 * grid}px`,

    globalNavContentHeight: `${4 * grid}px`,

    maxPageWidth: '1320px',
    minPageWidth: '1080px',
    pagePadding: '20px'
  },

  rawSizes: {
    grid,
    globalNavHeightRaw: 6 * grid,
    globalNavContentHeightRaw: 4 * grid,
    contextNavHeightRaw: 9 * grid
  },

  fonts: {
    baseFontFamily: "'Helvetica Neue', 'Segoe UI', Helvetica, Arial, sans-serif",
    systemFontFamily:
      "-apple-system,'BlinkMacSystemFont','Segoe UI','Helvetica','Arial',sans-serif",
    sonarcloudFontFamily:
      "Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif"
  },

  // z-index
  // =======
  //    1 -  100  for page elements (e.g. sidebars, panels)
  //  101 -  500  for generic page fixed elements (e.g. navigation, workspace)
  //  501 - 3000  for page ui elements
  // 3001 - 8000  for generic ui elements (e.g. dropdowns, tooltips)
  zIndexes: {
    // common
    aboveNormalZIndex: '3',
    normalZIndex: '2',
    belowNormalZIndex: '1',

    // ui elements
    pageMainZIndex: '50',

    tooltipZIndex: '8000',

    dropdownMenuZIndex: '7500',

    processContainerZIndex: '7000',

    modalZIndex: '6001',
    modalOverlayZIndex: '6000',

    popupZIndex: '5000'
  },

  others: {
    defaultShadow: '0 6px 12px rgba(0, 0, 0, 0.175)'
  }
};
