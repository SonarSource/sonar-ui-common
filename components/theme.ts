/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { css, ThemeContext as EmotionThemeContext } from '@emotion/core';
import emotionStyled, { CreateStyled } from '@emotion/styled';
import {
  ThemeProvider as EmotionThemeProvider,
  ThemeProviderProps,
  useTheme as emotionUseTheme,
  withTheme
} from 'emotion-theming';
import * as React from 'react';
// mockedTheme is in fact the default theme
// eslint-disable-next-line jest/no-mocks-import
import { mockedTheme } from './__mocks__/mockedTheme';

export interface Theme {
  colors: T.Dict<string>;
  sizes: T.Dict<string>;
  rawSizes: T.Dict<number>;
  fonts: T.Dict<string>;
  zIndexes: T.Dict<string>;
  others: T.Dict<string>;
}

export interface ThemedProps {
  theme: Theme;
}

const ThemeContext = EmotionThemeContext as React.Context<Theme>;

// Hack : override the default value of the context used for theme by emotion
// As we can't pass a default value when it is constructed
// This allows outside tests to get the mocked theme value without specifiying a theme provider
if (process.env.NODE_ENV === 'test') {
  (ThemeContext as any)['_currentValue'] = mockedTheme;
  (ThemeContext as any)['_currentValue2'] = mockedTheme;
}

export const styled = emotionStyled as CreateStyled<Theme>;
export const ThemeConsumer = ThemeContext.Consumer;
export const ThemeProvider = EmotionThemeProvider as React.ProviderExoticComponent<
  ThemeProviderProps<Theme>
>;
export const useTheme = emotionUseTheme as () => Theme;

export { css, withTheme };
export default ThemeContext;
