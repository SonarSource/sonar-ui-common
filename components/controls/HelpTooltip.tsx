/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import HelpIcon from '../icons/HelpIcon';
import { IconProps } from '../icons/Icon';
import { ThemeConsumer } from '../theme';
import './HelpTooltip.css';
import Tooltip, { Placement } from './Tooltip';

interface Props extends Pick<IconProps, 'size'> {
  className?: string;
  children?: React.ReactNode;
  onShow?: () => void;
  overlay: React.ReactNode;
  placement?: Placement;
}

export default function HelpTooltip({ size = 12, ...props }: Props) {
  return (
    <div className={classNames('help-tooltip', props.className)}>
      <Tooltip
        mouseLeaveDelay={0.25}
        onShow={props.onShow}
        overlay={props.overlay}
        placement={props.placement}>
        <span className="display-inline-flex-center">
          {props.children || (
            <ThemeConsumer>
              {(theme) => <HelpIcon fill={theme.colors.gray71} size={size} />}
            </ThemeConsumer>
          )}
        </span>
      </Tooltip>
    </div>
  );
}

export function DarkHelpTooltip({ size = 12, ...props }: Omit<Props, 'children'>) {
  return (
    <HelpTooltip {...props}>
      <ThemeConsumer>
        {({ colors }) => (
          <HelpIcon fill={colors.transparentBlack} fillInner={colors.white} size={size} />
        )}
      </ThemeConsumer>
    </HelpTooltip>
  );
}
