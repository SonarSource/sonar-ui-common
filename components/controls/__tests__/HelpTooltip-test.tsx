/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { ThemeProvider } from '../../theme';
// eslint-disable-next-line jest/no-mocks-import
import { mockedTheme } from '../../__mocks__/mockedTheme';
import HelpTooltip from '../HelpTooltip';

it('should render properly', () => {
  const wrapper = shallow(<HelpTooltip overlay={<div className="my-overlay" />} />, {
    wrappingComponent: ThemeProvider,
    wrappingComponentProps: {
      theme: mockedTheme,
    },
  });
  expect(wrapper).toMatchSnapshot();
});
