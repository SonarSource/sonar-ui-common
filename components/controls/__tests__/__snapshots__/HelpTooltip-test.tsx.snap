// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<div
  className="help-tooltip"
>
  <Tooltip
    mouseLeaveDelay={0.25}
    overlay={
      <div
        className="my-overlay"
      />
    }
  >
    <span
      className="display-inline-flex-center"
    >
      <ContextConsumer>
        <Component />
      </ContextConsumer>
    </span>
  </Tooltip>
</div>
`;
